docs/_docs/sql-reference/explain-help-with-performance-problems.adoc

```diff
@@ -0,0 +1,495 @@
+            table: PUBLIC.DEPT
+----
+We can see additional **Exchange** operator, which indicates that rows from one node are transferred into another.
+After changing 'lost' colocation:
+[source,sql]
+----
+-- explicitly colocated by (dept_id)
+CREATE TABLE emp(dept_id INTEGER, name VARCHAR, salary INTEGER, PRIMARY KEY(dept_id, name)) COLOCATE BY(dept_id);
+
+-- explicitly colocated by (id)
+CREATE TABLE dept(id INTEGER, name VARCHAR, PRIMARY KEY(id, name)) COLOCATE BY(id);
+
+SELECT emp.name, dept.name FROM emp JOIN dept ON emp.dept_id = dept.id;
+----
+Possible plan will show correct colocated join:
+[source,sql]
+----
+Exchange
+  Project
+    HashJoin
+      TableScan
+          table: PUBLIC.EMP
+      TableScan
+          table: PUBLIC.DEPT
+----
+== Optimized plans
+Some queries can be optimized to use more optimal plans which brings performance speed up. For example, plan for:
+[source,sql]
+----
+SELECT COUNT(*) FROM emp;
+----
+Can look like:
+[source,sql]
+----
+SelectCount
+    table: [PUBLIC, EMP]
+    est: (rows=43)
+----
+But there are numerous cases where such optimization is not applicable (transactional guarantees or some other restrictions), in such a case plan can be a bit different and more time consumable for execution.
+The same query as above, but with explicit transaction may produce a different plan like:
+[source,sql]
+----
+Project
+  ReduceSortAggregate
+    Exchange
+      MapSortAggregate
+        TableScan
+            table: PUBLIC.EMP
+            est: (rows=43)
+----
+== Index scan without exact search bounds
+Table scans are available in two implementations: direct table scan and scan through index, index scan containing 'predicate' and so-called: 'search bounds'.
+Predicate provides final rows comparison and in case when search bounds are absent it degenerates into table scan through index scan (additional store look up) with further predicate comparison which incurs additional performance overhead costs.
+Let`s suppose we have schema and query like:
+[source,sql]
+----
+CREATE TABLE t (id INTEGER PRIMARY KEY, col1 DECIMAL(5, 3));
+CREATE INDEX t_col1_idx ON t(col1);
+
+SELECT id FROM t WHERE col1 = 43::DECIMAL(5, 3);
+----
+And possible plan :
+[source,sql]
+----
+Exchange
+  IndexScan
+      table: PUBLIC.T
+      index: T_COL1_IDX
+      predicate: =(COL1, 43.000)
+      searchBounds: [ExactBounds [bound=43.000:DECIMAL(5, 3)]]
+----
+We can see here both **searchBounds** and **predicate** which means that only exact lookup through index will be involved.
+But without cast it looks different:
+[source,sql]
+----
+SELECT id FROM t WHERE col1 = 43;
+----
+[source,sql]
+----
+Exchange
+  IndexScan
+      table: PUBLIC.T
+      index: T_COL1_IDX
+      predicate: =(CAST(COL1):DECIMAL(13, 3), 43.000)
+----
+We can see here - only **predicate** which means that **all** rows from index will go through predicate and bring additional performance penalty.
+Such kind a plan is less preferable than a table full scan.
```

Comment:
I'd prefer to see optimal and non-optimal plan in the same order across the whole document.
So the structure could be either:
* typical query
* non-optimal plan
* short analysis and suggestion better query
* optimal plan
or (less preferable):
* optimal query
* common errors analysis